bazel_repositories_priv

```diff
@@ -0,0 +1,292 @@
+        if filename in archives:
+            continue
+        try:
+            cache_filename = process_url(url, url_path, filename)
+        except (urllib_error.HTTPError, urllib_error.URLError):
+            log.warning(f"could not download {url}")
+            continue
+        else:
+            archives.add(filename)
+            # If dependency is excluded, don't add it to the result list,
+            # remove the archive from the working dir, but keep it in the
+            # cache (otherwise next iterations of ´bazel cache` will not work).
+            for e in exclude:
+                if e in url:
+                    log.info(f"excluding dependency {url}")
+                    append_url = False
+            if append_url:
+                filtered_urls.append(url)
+            else:
+                excluded.append(cache_filename)
+    return filtered_urls, excluded
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description=description,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument("--archive")
+    parser.add_argument("--bazel-args")
+    parser.add_argument("--exclude")
+    parser.add_argument("--home")
+    parser.add_argument("--target")
+    parser.add_argument("--user")
+    args = parser.parse_args()
+
+    log.debug(f"target: {args.target}")
+
+    exclude = args.exclude.split(",") if args.exclude else []
+    root_dir = find_rootdir(args.archive)
+
+    if os.getuid() != 0:
+        exit("You have to be root")
+
+    global user
+    user = args.user
+
+    netns_id = uuid.uuid4()
+    netns_name = f"obs-bazel-{netns_id}"
+
+    subprocess.run(["ip", "netns", "add", netns_name])
+    subprocess.run(["ip", "netns", "exec", netns_name,
+                    "ip", "link", "set", "lo", "up"])
+
+    os.makedirs(CACHEDIR, exist_ok=True)
+
+    shutil.rmtree(os.path.join("/home", args.user, ".cache", "bazel"),
+                  ignore_errors=True)
+    p = multiprocessing.Process(target=bazel_clean, args=(root_dir,))
+    p.start()
+    p.join()
+
+    all_urls = []
+    all_excluded = []
+    parent_conn, child_conn = multiprocessing.Pipe()
+    while True:
+        log.info("Running Bazel and checking for dependencies to download")
+        p = multiprocessing.Process(target=sub, args=(child_conn, netns_name,
+                                                      root_dir,
+                                                      args.bazel_args,
+                                                      args.target,))
+        p.start()
+        returncode, urls = parent_conn.recv()
+        p.join()
+        if returncode == 0:
+            break
+        log.debug(f"urls: {urls}")
+        new_urls, new_excluded = process_urls(urls, exclude)
+        all_urls += new_urls
+        all_excluded += new_excluded
+
+    for e in all_excluded:
+        os.unlink(e)
+    chown(ROOT_CACHEDIR)
+
+    subprocess.run(["ip", "netns", "del", netns_name])
```

Comment:
Should this maybe be included in an atexit handler so it'll still run in the event of encountering an exception elsewhere?
https://docs.python.org/3/library/atexit.html

Perhaps in a future version.

Reply:
Good idea, I will try now.